x-pack/plugins/viz_editor/common/datasource.ts

```diff
@@ -4,9 +4,21 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+import { Scale } from '../public/common/components/operation_editor';
+
+export type FieldType = 'number' | 'string' | 'date';
```

Comment:
Actually I would like to simplify the `types` our editors reason about as much as possible - there is definitely a point for boolean, geo_point, geo_shape and ip, but I'm not sure whether it makes sense to expose stuff like `conflict` or `murmur3` to a visualization tool. If a data source wants to work with these things under the hood, it has to abstract that away, e.g. by mapping it to multiple virtual fields and keeping track of this to fetch the right data. IMHO we shouldn't have dynamic types of fields because it makes the interaction between the various editor parts much more complicated.

Maybe `type` isn't a useful abstraction here at all and we should be talking about the scale type with `nominal`, `ordinal`, `interval` and `ratio` instead.

cc ing @timroes and @chrisdavies here because this is definitely a topic open for discussion.

Reply:
I hear what you're saying, but I don't think your answer handles the main issue in my previous comment. Specifically, our current index patterns already support all of the types listed above. I can see two options for how to handle this:

- Modify the data loader in `index_patterns.ts` to map to these types- otherwise you will have static types that don't match runtime
- Use a string type and do this all at runtime, preferably reusing some existing logic such as `kbn_field_types.js`

Basically my concern is that these type definitions will be incorrect and misleading if we use them as-is.

Reply:
> Modify the data loader in index_patterns.ts to map to these types- otherwise you will have static types that don't match runtime

Yes, this is exactly what I had in mind, but preferably in a separate PR as this one got big enough. My concrete suggestion for handling this for index patterns is to map all types which don't make sense for us to strings and just specify in the field capabilities (which don't really exist in the moment aside from sortable, filterable and aggregatable) that you can only show them as a naked value and nothing else - or even filtering them out completely because they are just noise.

Maybe we are having different things in mind when speaking about types - if this is a random string coming from a dynamic registry, editor plugins can't really act on it, e.g. in terms of deciding whether it makes sense to be droppable in a certain spot or not. We need some specified way to decide whether a field can or can't be used in a specific context, and at the moment this is the `type` of the field together with `aggregatable`.

I think there are multiple possible solutions for that:
1. We decide what to do with a field by `type` (which gives the basic information which is enough for a lot of use cases) and `aggregatable`, to be extended by `supportedOperations` or something (to cover things like rollup)
2. `type` is just a string which in itself doesn't tell the editor plugin anything (maybe should be called `rawType` or `underlyingType`), but there are other properties to decide on like `scaleType`: nominal, ordinal, interval and ratio and `cardinality`: number
3. A field has a function which the editor plugin can call to get informations about what this field is and cares about (I'm not sure how different this is from number 2)

Reply:
@flash1293 I'd prefer discussing this in another PR- in this PR, please don't start moving towards the restricted types you've done in this file